docs/plugin-recipes/discord-bot-key-tutorial.md

```diff
@@ -0,0 +1,173 @@
+
+```typescript
+if (dngServer.length > 0) {
+    bot.channels.get('channel-id').send(`${dngName} opened by ${dngOpener} in ${dngServer}`)
+}
+```
+You will want to reset the dungeon variables back to being empty (`''`) so it won't resend the same dungeon.
+Don't forget to add the bot to your server!
+
+The final file should look like this:
+```typescript
+// These 5 imports are essential for any plugin.
+import { NrPlugin, HookPacket, Packet, PacketType, Client} from './../core/plugin-module';
+
+// Import the TextPacket
+import { TextPacket } from '../networking/packets/incoming/text-packet';
+
+// Define the dungeon variables.
+let dngName = '';
+let dngOpener = '';
+let dngServer = '';
+
+// Create a discord bot
+const Discord = require('discord.js');
+const bot = new Discord.Client();
+
+// Add the token
+bot.login('token');
+
+// Run the bot
+bot.on('ready', () => {
+    setInterval(function() {
+        // Check if there is a dungeon opened
+        if (dngServer.length > 0) {
+            bot.channels.get('channel-id').send(`${dngName} opened by ${dngOpener} in ${dngServer}`)
+            dngName = '';
+            dngOpener = '';
+            dngServer = '';
+        }
+    }, 500);
+})
+// The NrPlugin decorator gives nrelay some information about
+// your plugin. If it is not present, nrelay won't load the plugin.
+@NrPlugin({
+    name: 'Key Notifier Plugin',
+    author: 'DavisXola'
+})
+class KeyPlugin {
+
+    private dngRegex = /^{"key":"server.dungeon_opened_by","tokens":{"dungeon":"(\S.*)", "name":"(\w+)"}}$/;
+
+    // The HookPacket decorator will cause the method to be called
+    // whenever a packet with the specified packet type is recieved.
+    @HookPacket(PacketType.TEXT)
+    // Any method with a HookPacket decorator should always have
+    // the method signature (client: Client, packet: Packet).
+    onTextPacket(client: Client, textPacket: TextPacket): void {
+
+        // Check if a key is opened
+        const match = this.dngRegex.exec(textPacket.text);
+        if (match) {
+            dngName = match[1];
+            dngOpener = match[2];
+            dngServer = client.server.name;
+        }
+    }
+}
+```
+I also made it so the message gets deleted after 30 seconds (when the portal is closed) and 5 seconds before that, it edits the message to have a "(closing)" at the end.
+You can add that by fetching the message upon sending it and using `.then` to edit it and delete it.
+```typescript
+bot.channels.get('426565182718738445').send(`${ping} ${dngName} opened by ${dngOpener} in ${dngServer}`)
+    .then(function (msg) {
+        setTimeout(() => {
+            msg.edit(msg.content + ` (closing)`);
+        }, 25000);
+        setTimeout(() => {
+            msg.delete();
+        }, 30000);
+    })
+```
+**Warning: Using the `.then` and getting the message can only work if you change `noImplicitAny` from `true` to `false` in `tsconfig.json`**
```

Comment:
Why don't you just specify the type of `msg` explicitly to fix the error instead of changing the config?

Reply:
I tried to specify the type `Message` which is the correct type but it didn't recognize it for some reason.

Reply:
Might be a missing import or something?

Reply:
I thought about it but on my discord bots that are coded in JS the import I need is discord itself, which I did import. So I can't seem to know what's missing